site/docs/expressions/Extension_Functions/index.md

```diff
@@ -0,0 +1,219 @@
+
+Note that this imposes some constraints on the allowed patterns. For example, you can't define a function like `cast(T) -> S` because `S` is not yet defined and is therefore ambiguous. Most patterns have well-defined semantics in both match and evaluation context, though.
+
+### Expressions
+
+For some functions, such as those operating on `FIXEDCHAR` or `DECIMAL` types, we may need to do some math to derive the return type. Here's a simple example: `concatenate(FIXEDCHAR<A>, FIXEDCHAR<B>) -> FIXEDCHAR<A + B>`.
+
+To keep things simple, Substrait defines only a basic set of operators, functions, and metatypes for type derivations, and this set of functions is not extensible. The complete list can be found [here](../../types/meta_type_system/#functions).
+
+!!! note
+
+    These functions are not to be confused with extension functions or embedded functions, or anything else that exists when a plan is being executed. They exist on a completely different abstraction layer! In fact, we need these functions in order to be able to even begin thinking about defining the more complex functions that we use in plans, and we can't use functions to define themselves.
+
+## Derivation Programs
+
+For more advanced functions, like the decimal addition we used as an example earlier, a single pattern is just not enough to derive the return type. For this purpose, the return type can include a number of statements before the final pattern, separated from each other and the pattern by newlines or semicolons. These statements more or less look and behave like assignment statements, except that the left-hand side can be any pattern, and the assignment utilizes the same matching process used for arguments.
+
+!!! warning
+
+    The semantics for matching mean that the names used in these assignment statements don't act like variables when they are reused! For example, `bad_concatenate(FIXEDCHAR<A>, FIXEDCHAR<B>) -> A = A + B; FIXEDCHAR<A>` will *not* behave like the `concatenate` example from before, because unless `B` equals 0, we can't reassign `A` to a different value.
+
+This simple rule not only lets us eliminate common subexpressions to make the decimal `add` somewhat comprehensible, but it also lets us write essentially arbitrary constraints. For example, we could define an integer addition that works with any combination of integers as follows:
+
+```
+casting_add(LHS, RHS, type RETURN) ->
+  true = LHS == i8 || LHS == i16 || LHS == i32 || LHS == i64
+  true = RHS == i8 || RHS == i16 || RHS == i32 || RHS == i64
+  true = RETURN == i8 || RETURN == i16 || RETURN == i32 || RETURN == i64
+  RETURN
+```
```

Comment:
bad

Reply:
F

Any suggestions beyond the `assert` statement? I considered at some point to introduce a pattern union operator that would in this case reduce the whole thing down to just

```
casting_add(i8 | i16 | i32 | i64, i8 | i16 | i32 | i64, type T) ->
  assert covers(T, i8 | i16 | i32 | i64)
  T
```

or something (the equalities in the original can also be replaced with `covers`, but I wasn't sure if I wanted to get into that function on the overview page), but I didn't want to make the whole thing more complicated than it already is when it doesn't really add to the expressiveness.

Reply:
Or with `&` too: `casting_add(i8 | i16 | i32 | i64, i8 | i16 | i32 | i64, type T & i8 | i16 | i32 | i64) -> T`. I would've liked `:` in that position instead, but it's a pretty overloaded character already... Could also use `{i8, i16, i32, i64}` instead of defining the `|` operator, so `|` and `&` can be reserved for bitwise integer operations.